### ISSUE TITLE
如何將 API 回傳的 Story Format 資料，轉換成 Plugin 的 story_data & html

### ISSUE THREAD

**Issue by lucemia** [livingbio/web-stories-wp#8]
relate to: #4

# Problem
web-stories-wp 需要自己的 story_data (editor 用的格式) & amp html (顯示用)
需要設計一個機制將 gstudio 新 API https://github.com/livingbio/gstudio/pull/1857 回傳的 JSON 格式，分別轉換成 web-stories-wp 的 JSON 格式 & 轉換成最後的 AMP HTML

ref: https://github.com/livingbio/web-stories-wp/issues/2

`GliaStudio's Story Format -> Converter (+ Template) -> web-storeis-wp JSON -> AMP Html`

# Solution

因為目前 Template / convert to AMP HTML 的機制都在 Frontend (React)，但 DB 控制都在 Backend
參考 google#2140 的做法，寫成一個 npm workflow 的方式，讓 Backend call cli 來轉換，後存到 DB

`npm run workflow:aiconvert -i input.json story-tmp.json story-tmp.html`

# Action
- [x] Provide Sample Data @lucemia

**Comment by lucemia** [comment]
template 中有 background 的欄位，但沒有 title, description 欄位
https://github.com/livingbio/web-stories-wp/tree/master/assets/src/dashboard/templates/data
看要手動改 Template & 另外附加檔案處理

**Comment by lucemia** [comment]
Get Story Data -> Transfer to `story_data` -> Transfer to html

**Comment by lucemia** [comment]
參考 
https://github.com/google/web-stories-wp/pull/2140

`npm run workflow:templates`

**Comment by vickyliin** [comment]
# State Management of Story

所有資料都即時在前端，只有 `save` & `autoSave` 時才會有 API 交換。

**Usage**

```javascript
import { useStory } from './edit-story/story'

const { state, actions } = useStory()

// modify state
actions[apiAction](payload)
```

### State

- States defined in `storyProvider.js`
https://github.com/livingbio/web-stories-wp/blob/f29c49d19a4c8c6fee071710a3767610102c8690/assets/src/edit-story/app/story/storyProvider.js#L113-L125

- Some states defined in `useStoryReducer.js`
https://github.com/livingbio/web-stories-wp/blob/f29c49d19a4c8c6fee071710a3767610102c8690/assets/src/edit-story/app/story/useStoryReducer/useStoryReducer.js#L28-L33


### Actions
- `apiActions` defined in [actions.js](https://github.com/livingbio/web-stories-wp/blob/dev/assets/src/edit-story/app/story/useStoryReducer/actions.js)
- Implementations defined in [reducer.js](https://github.com/livingbio/web-stories-wp/blob/dev/assets/src/edit-story/app/story/useStoryReducer/reducer.js)


## Trace

**Prior**

- [React.useContext](https://zh-hant.reactjs.org/docs/hooks-reference.html#usecontext)
- [React.useCallback](https://zh-hant.reactjs.org/docs/hooks-reference.html#usecallback)
- [use-context-selector](https://github.com/dai-shi/use-context-selector)
- [react-pure-render: shallowEqual](https://github.com/gaearon/react-pure-render#shallowequal)


**Entry**
- [storyProvider.js](https://github.com/livingbio/web-stories-wp/blob/dev/assets/src/edit-story/app/story/storyProvider.js)
- [useStory.js](https://github.com/livingbio/web-stories-wp/blob/dev/assets/src/edit-story/app/story/useStory.js)

**Shared**: `Context`
https://github.com/livingbio/web-stories-wp/blob/f29c49d19a4c8c6fee071710a3767610102c8690/assets/src/edit-story/app/story/context.js#L22

### [storyProvider.js](https://github.com/livingbio/web-stories-wp/blob/dev/assets/src/edit-story/app/story/storyProvider.js)

Provide `state` & `actions` from [useStoryReducer](https://github.com/livingbio/web-stories-wp/blob/dev/assets/src/edit-story/app/story/useStoryReducer/useStoryReducer.js)

https://github.com/livingbio/web-stories-wp/blob/f29c49d19a4c8c6fee071710a3767610102c8690/assets/src/edit-story/app/story/storyProvider.js#L111-L133


### [useStoryReducer](https://github.com/livingbio/web-stories-wp/blob/dev/assets/src/edit-story/app/story/useStoryReducer/useStoryReducer.js)


```javascript
const { state, internal, api } = useStoryReducer()

// set state
api[actionName](payload)
internal[actionName](payload)
```

- Actions defined in [actions.js](https://github.com/livingbio/web-stories-wp/blob/dev/assets/src/edit-story/app/story/useStoryReducer/actions.js)
- Implementations defined in [reducer.js](https://github.com/livingbio/web-stories-wp/blob/dev/assets/src/edit-story/app/story/useStoryReducer/reducer.js)


### [utils/context.js](https://github.com/livingbio/web-stories-wp/blob/dev/assets/src/edit-story/utils/context.js)

和 [use-context-selector](https://github.com/dai-shi/use-context-selector) 相同用法，但是 selector 結果如果 shallowEqual 上次的結果，就回傳上次的結果。

**Comment by vickyliin** [comment]
@lucemia 初步判斷可以寫成 cli 型式，但是沒辦法像 https://github.com/google/web-stories-wp/pull/2140 一樣直接寫在 bin/utils 裡。要在 `webpack.config.cjs` 新增 entrypoint, 設定 compile 成 server side javascript，之後再執行 compile 過的 script，這樣 ok 嗎？

**Comment by lucemia** [comment]
> @lucemia 初步判斷可以寫成 cli 型式，但是沒辦法像 [google#2140](https://github.com/google/web-stories-wp/pull/2140) 一樣直接寫在 bin/utils 裡。要在 `webpack.config.cjs` 新增 entrypoint, 設定 compile 成 server side javascript，之後再執行 compile 過的 script，這樣 ok 嗎？

可以的!

**Comment by vickyliin** [comment]
試 compile 了一下發現整個 compile 進太多東西，裡面有些是瀏覽器才能使用，試試 tree shaking 能不能解決這問題，如果不行可能就只能用開網頁的方式了

**Comment by lucemia** [comment]
瀏覽器的部分，有可能像 jest 一樣透過 jsdom 解決嗎?

> The function being tested adds an event listener on the #button DOM element, so we need to set up our DOM correctly for the test. Jest ships with jsdom which simulates a DOM environment as if you were in the browser. This means that every DOM API that we call can be observed in the same way it would be observed in a browser!

https://dev.to/ms314006/use-jest-write-unit-testing-for-dom-manipulation-3n6c

確定不行 / 或太困難的話和我說一聲，我考慮看看要不要換流程 ~

**Comment by vickyliin** [comment]
用 jsdom 測試可以成功把 `getStoryMarkup` import 進來了！我再測測看能不能用

**Comment by vickyliin** [comment]
基本的 `renderToStaticMarkup` 可行:

```javascript
import { renderToStaticMarkup } from 'react-dom/server';

function MyComponent() {
  return <div>{'Hello World'}</div>;
}

export function aiconvert() {
  console.log(renderToStaticMarkup(<MyComponent />));
}
```

```sh
$ npm run workflow:aiconvert

> web-stories-wp@ workflow:aiconvert
> ./bin/build/commander.js aiconvert

<div>Hello World</div>
```

**Comment by lucemia** [comment]
> 用 jsdom 測試可以成功把 `getStoryMarkup` import 進來了！我再測測看能不能用

太厲害了!!

**Comment by vickyliin** [comment]
無法 push:
```refusing to allow an OAuth App to create or update workflow `.github/workflows/continuous-integration-build.yml` without `workflow` scope```

不過沒有做這些修改耶，還是因為 workflow:aiconvert ?

**Comment by lucemia** [comment]
> 無法 push:
> `` refusing to allow an OAuth App to create or update workflow `.github/workflows/continuous-integration-build.yml` without `workflow` scope ``
> 
> 不過沒有做這些修改耶，還是因為 workflow:aiconvert ?

明天來看看 ..

**Comment by lucemia** [comment]
https://github.community/t/refusing-to-allow-an-integration-to-create-or-update/16326/2

**Comment by lucemia** [comment]
fixed in #23
